scheds/include/scx/sdt_task_impl.bpf.h

```diff
@@ -0,0 +1,482 @@
+
+		desc->nr_free += 1;
+		if (desc->nr_free > 1)
+			break;
+	}
+
+	bpf_spin_unlock(&sdt_task_lock);
+	return;
+}
+
+static SDT_TASK_FN_ATTRS void sdt_task_free(struct task_struct *p)
+{
+	struct sdt_task_map_val *mval;
+
+	sdt_arena_verify();
+
+	mval = bpf_task_storage_get(&sdt_task_map, p, 0, 0);
+	if (!mval)
+		return;
+
+	sdt_task_free_idx(mval->tid.idx);
+	mval->data = NULL;
+}
+
+
+static SDT_TASK_FN_ATTRS int sdt_task_find_empty(struct sdt_task_desc __arena *desc, struct sdt_task_desc * __arena *descp, __u64 *idxp)
+{
+	struct sdt_task_desc * __arena *desc_children,  __arena *new_chunk;
+	struct sdt_task_desc __arena *lv_desc[SDT_TASK_LEVELS];
+	struct sdt_task_chunk __arena *chunk;
+	struct sdt_task_desc __arena *tmp;
+	__u64 lv_pos[SDT_TASK_LEVELS];
+	__u64 u, pos, level;
+	__u64 idx = 0;
+
+	bpf_for(level, 0, SDT_TASK_LEVELS) {
+		pos = sdt_chunk_find_empty(desc);
+		if (pos < 0)
+			return -ENOMEM;
+
+		idx <<= SDT_TASK_ENTS_PER_CHUNK_SHIFT;
+		idx |= pos;
+
+		/* Log the levels to complete allocation. */
+		lv_desc[level] = desc;
+		lv_pos[level] = pos;
+
+		/* The rest of the loop is for internal node traversal. */
+		if (level == SDT_TASK_LEVELS - 1)
+			break;
+
+		cast_kern(desc);
+
+		chunk = desc->chunk;
+
+		cast_kern(chunk);
+
+		desc_children = (struct sdt_task_desc * __arena *)chunk->descs;
+		desc = desc_children[pos];
+
+		/* Someone else is populating the subtree. */
+		if (desc == (void *)SDT_TASK_ALLOC_RESERVE)
+			return -EAGAIN;
+
+		if (!desc) {
+			/* Reserve our spot and go allocate. */
+			desc_children[pos] = (void *)SDT_TASK_ALLOC_RESERVE;
+
+			bpf_spin_unlock(&sdt_task_lock);
+			new_chunk = sdt_alloc_chunk();
+			bpf_spin_lock(&sdt_task_lock);
+
+			if (!new_chunk)
+				return -ENOMEM;
+
+			desc_children[pos] = new_chunk;
+		}
+	}
+
+	bpf_for(u, 0, SDT_TASK_LEVELS) {
+		level = SDT_TASK_LEVELS - 1 - u;
+		tmp = lv_desc[level];
+		cast_kern(tmp);
+
+		sdt_set_idx_state(tmp, lv_pos[level], true);
+
+		tmp->nr_free -= 1;
+		if (tmp->nr_free > 0)
+			break;
+	}
+
+	*descp = desc;
+	*idxp = idx;
+
+	return 0;
+}
+
+static SDT_TASK_FN_ATTRS struct sdt_task_data __arena *sdt_task_alloc(struct task_struct *p)
+{
+	struct sdt_task_data __arena *data = NULL;
+	struct sdt_task_desc __arena *desc;
+	struct sdt_task_chunk __arena *chunk;
+	struct sdt_task_map_val *mval;
+	__u64 idx, pos;
+	int ret;
+
+	mval = bpf_task_storage_get(&sdt_task_map, p, 0,
+				    BPF_LOCAL_STORAGE_GET_F_CREATE);
+	if (!mval)
+		return NULL;
+
+	bpf_spin_lock(&sdt_task_lock);
+
+	bpf_repeat(SDT_TASK_ALLOCATION_ATTEMPTS) {
+		ret = sdt_task_find_empty(sdt_task_desc_root, &desc, &idx);
+		if (ret != -EAGAIN)
+			break;
+	}
+
+	if (ret != 0) {
+		bpf_spin_unlock(&sdt_task_lock);
+		return NULL;
+	}
+
+	cast_kern(desc);
+
+	chunk = desc->chunk;
+	cast_kern(chunk);
+
+	/* populate leaf node if necessary */
+
+	pos = idx & (SDT_TASK_ENTS_PER_CHUNK - 1);
+	data = chunk->data[pos];
+	if (!data) {
+		bpf_spin_unlock(&sdt_task_lock);
```

Comment:
I guess `sdt_task_alloc_from_pool` needs the lock as well. I think that should be fine.

Reply:
```sdt_task_alloc_from_pool``` uses the ```sdt_pool_alloc_lock```, so we cannot keep the task lock even if we use the fast path in the pool allocator since IIRC we cannot hold multiple spinlocks. The other reason is that on the slow path we have to use the ```bpf_arena_alloc_pages``` kfunc so we cannot be holding any locks.